src/codegen/compile.ml

```diff
@@ -8187,18 +8209,34 @@ let rec compile_lexp (env : E.t) ae lexp =
   match lexp.it with
   | VarLE var -> Var.set_val env ae var
   | IdxLE (e1, e2) ->
-     compile_exp_vanilla env ae e1 ^^ (* offset to array *)
-     compile_exp_vanilla env ae e2 ^^ (* idx *)
-     Arr.idx_bigint env,
-     SR.Vanilla,
-     store_ptr
+    let (set_field, get_field) = new_local env "field" in
+    compile_exp_vanilla env ae e1 ^^ (* offset to array *)
+    compile_exp_vanilla env ae e2 ^^ (* idx *)
+    Arr.idx_bigint env ^^
+    set_field ^^
+    get_field,
+    SR.Vanilla,
+    store_ptr ^^
+    (if !Flags.gc_strategy = Mo_config.Flags.Generational
+     then
+      get_field ^^
+      compile_add_const ptr_unskew ^^
+      E.call_import env "rts" "write_barrier"
+     else G.nop)
   | DotLE (e, n) ->
-     compile_exp_vanilla env ae e ^^
-     (* Only real objects have mutable fields, no need to branch on the tag *)
-     Object.idx env e.note.Note.typ n,
-     SR.Vanilla,
-     store_ptr
-
+    let (set_field, get_field) = new_local env "field" in
+    compile_exp_vanilla env ae e ^^
+    Object.idx env e.note.Note.typ n ^^
+    set_field ^^
+    get_field,
+    SR.Vanilla,
+    store_ptr ^^
+    (if !Flags.gc_strategy = Mo_config.Flags.Generational
+    then  
+      get_field ^^
+      compile_add_const ptr_unskew ^^
+      E.call_import env "rts" "write_barrier"
+    else G.nop)
```

Comment:
I guess in all these cases the local "field" is only necessary if we need to insert the write barrier. 
It might be better to avoid declaring and storing the field when the write barriers are not needed, perhaps
by just having two version of this function and switching appropriately (or using Ocaml patttern match guards on the cases, switching on the compiler flag).

Reply:
BTW, what happened to the type based optimization (no write barriers for always scalar types)? Perhaps best in a follow on PR, but I thought you'd already done it!)

Reply:
Also, wasm has tee instruction that sets a local without popping the value (which might useful) to save on get_field

Reply:
I will improve this with separate versions.

Reply:
I have implemented the static optimization as preparation for the incremental GC and thought to include it later as a separate PR (to reduce complexity and risk).

Reply:
I added your PR. Thanks again.